PR: https://github.com/odoo/odoo/pull/78287

From: 47041f2d45915e2ed74c4317f9aafe3dfa840f0a
From: Fabio Barbero
Date: 2022-03-14 08:42:58

Structural Changes: 4
Total Changes: 125

[IMP] base: allow user to activate multiple languages at once

Purpose
=======
Add "Activate" button when selecting multiple languages, select multiple
languages when clicking on "Add languages" in settings.

Specifications
=============
`lang` variable in `base.language.install` changed from Selection to
Many2many to allow multiple languages being activated at once.
Hide globe icon for language fields from view mode (only visible when
editing). Remove state in base.language.install since it's no longer
need to keep track of the installation step.

Task-2662548

closes odoo/odoo#78287

Related: odoo/upgrade#2921
Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/odoo/addons/base/models/res_lang.py
+++ b/odoo/addons/base/models/res_lang.py
@@ -334,6 +334,22 @@ class Lang(models.Model):
 
         return formatted
 
+    def action_activate_langs(self):
+        """ Activate the selected languages """
+        for lang in self.filtered(lambda l: not l.active):
+            lang.toggle_active()
+        message = _("The languages that you selected have been successfully installed. Users can choose their favorite language in their preferences.")
+        return {
+            'type': 'ir.actions.client',
+            'tag': 'display_notification',
+            'target': 'new',
+            'params': {
+                'message': message,
+                'type': 'success',
+                'sticky': False,
+                'next': {'type': 'ir.actions.act_window_close'},
+            }
+        }
 
 def split(l, counts):
     """

--- a/odoo/addons/base/views/res_lang_views.xml
+++ b/odoo/addons/base/views/res_lang_views.xml
@@ -6,6 +6,9 @@
             <field name="model">res.lang</field>
             <field name="arch" type="xml">
                 <tree string="Languages" limit="200">
+                    <header>
+                        <button name="action_activate_langs" type="object" string="Activate"/>
+                    </header>
                     <field name="name"/>
                     <field name="code" groups="base.group_no_one"/>
                     <field name="iso_code" groups="base.group_no_one"/>

--- a/odoo/addons/base/views/res_partner_views.xml
+++ b/odoo/addons/base/views/res_partner_views.xml
@@ -223,7 +223,7 @@
                                 <button
                                     type="action"
                                     name="%(base.res_lang_act_window)d"
-                                    class="btn-sm btn-link mb4 fa fa-globe"
+                                    class="oe_edit_only btn-sm btn-link mb4 fa fa-globe"
                                     aria-label="More languages"
                                     groups="base.group_system"
                                     title="More languages"

--- a/odoo/addons/base/views/res_users_views.xml
+++ b/odoo/addons/base/views/res_users_views.xml
@@ -245,7 +245,7 @@
                                             <button
                                                 type="action"
                                                 name="%(base.res_lang_act_window)d"
-                                                class="btn-sm btn-link mb4 fa fa-globe"
+                                                class="oe_edit_only btn-sm btn-link mb4 fa fa-globe"
                                                 aria-label="More languages"
                                                 title="More languages"/>
                                         </div>
@@ -447,7 +447,7 @@
                                         <button
                                             type="action"
                                             name="%(base.res_lang_act_window)d"
-                                            class="btn-sm btn-link mb4 fa fa-globe"
+                                            class="oe_edit_only btn-sm btn-link mb4 fa fa-globe"
                                             aria-label="More languages"
                                             groups="base.group_system"
                                             title="More languages"

--- a/odoo/addons/base/wizard/base_language_install.py
+++ b/odoo/addons/base/wizard/base_language_install.py
@@ -9,45 +9,61 @@ class BaseLanguageInstall(models.TransientModel):
     _description = "Install Language"
 
     @api.model
-    def _default_language(self):
+    def _default_lang_ids(self):
         """ Display the selected language when using the 'Update Terms' action
             from the language list view
         """
         if self._context.get('active_model') == 'res.lang':
-            lang = self.env['res.lang'].browse(self._context.get('active_id'))
-            return lang.code
+            return self._context.get('active_ids') or [self._context.get('active_id')]
         return False
 
-    @api.model
-    def _get_languages(self):
-        return [[code, name] for code, _, name, *_ in self.env['res.lang'].get_available()]
-
-    lang = fields.Selection(_get_languages, string='Language', required=True,
-                            default=_default_language)
+    # add a context on the field itself, to be sure even inactive langs are displayed
+    lang_ids = fields.Many2many('res.lang', 'res_lang_install_rel',
+                                'language_wizard_id', 'lang_id', 'Languages',
+                                default=_default_lang_ids, context={'active_test': False})
     overwrite = fields.Boolean('Overwrite Existing Terms',
                                default=True,
                                help="If you check this box, your customized translations will be overwritten and replaced by the official ones.")
-    state = fields.Selection([('init', 'init'), ('done', 'done')],
-                             string='Status', readonly=True, default='init')
+    first_lang_id = fields.Many2one('res.lang',
+                                    compute='_compute_first_lang_id',
+                                    help="Used when the user only selects one language and is given the option to switch to it")
+
+    def _compute_first_lang_id(self):
+        self.first_lang_id = False
+        for lang_installer in self.filtered('lang_ids'):
+            lang_installer.first_lang_id = lang_installer.lang_ids[0]
 
     def lang_install(self):
         self.ensure_one()
         mods = self.env['ir.module.module'].search([('state', '=', 'installed')])
-        self.env['res.lang']._activate_lang(self.lang)
-        mods._update_translations(self.lang, self.overwrite)
-        self.state = 'done'
+        lang_ids = self.lang_ids
+        langs_to_activate = lang_ids.filtered(lambda l: not l.active)
+        langs_to_activate.toggle_active()
+        mods._update_translations(lang_ids.mapped('code'), self.overwrite)
         self.env.cr.execute('ANALYZE ir_translation')
 
+        if len(lang_ids) == 1:
+            return {
+                'type': 'ir.actions.act_window',
+                'res_model': 'base.language.install',
+                'res_id': self.id,
+                'view_mode': 'form',
+                'target': 'new',
+                'views': [[self.env.ref('base.language_install_view_form_lang_switch').id, 'form']],
+            }
+
         return {
-            'name': _('Language Pack'),
-            'view_mode': 'form',
-            'view_id': False,
-            'res_model': 'base.language.install',
-            'domain': [],
+            'type': 'ir.actions.client',
+            'tag': 'display_notification',
             'context': dict(self._context, active_ids=self.ids),
-            'type': 'ir.actions.act_window',
             'target': 'new',
-            'res_id': self.id,
+            'params': {
+                'message': _("The languages that you selected have been successfully installed.\
+                            Users can choose their favorite language in their preferences."),
+                'type': 'success',
+                'sticky': False,
+                'next': {'type': 'ir.actions.act_window_close'},
+            }
         }
 
     def reload(self):
@@ -57,7 +73,7 @@ class BaseLanguageInstall(models.TransientModel):
         }
 
     def switch_lang(self):
-        self.env.user.lang = self.lang
+        self.env.user.lang = self.first_lang_id.code
         return {
             'type': 'ir.actions.client',
             'tag': 'reload_context',

--- a/odoo/addons/base/wizard/base_language_install_views.xml
+++ b/odoo/addons/base/wizard/base_language_install_views.xml
@@ -2,34 +2,48 @@
 <odoo>
     <data>
 
+        <record id="base.language_install_view_form_lang_switch" model="ir.ui.view">
+            <field name="name">Switch to language</field>
+            <field name="model">base.language.install</field>
+            <field name="priority">100</field>
+            <field name="arch" type="xml">
+                <form string="Switch to language">
+                    <group>
+                        <span class="o_form_label">
+                            <strong><field name="first_lang_id" readonly="True" options="{'no_open': True}"/></strong> has been successfully installed.
+                            Users can choose their favorite language in their preferences.
+                        </span>
+                    </group>
+                    <footer>
+                        <button name="reload" string="Close" type="object" class="btn-primary" data-hotkey="q"/>
+                        <button name="switch_lang" type="object" class="btn-primary ml-1" data-hotkey="w">
+                            Switch to <field name="first_lang_id" class="text-white" style="pointer-events: none;"/> &amp; Close
+                        </button>
+                    </footer>
+                </form>
+           </field>
+        </record>
+
         <record id="view_base_language_install" model="ir.ui.view">
             <field name="name">Load a Translation</field>
             <field name="model">base.language.install</field>
             <field name="arch" type="xml">
                 <form string="Load a Translation">
-                    <field name="state" invisible="1"/>
-                    <group states="init">
-                        <field name="lang"/>
+                    <group>
+                        <field name="lang_ids" widget="many2many_tags"
+                            context="{'active_test': False}" options="{'no_quick_create': True, 'no_create_edit': True}"/>
                         <field name="overwrite" groups="base.group_no_one"/>
                     </group>
-                    <group states="done" colspan="4">
-                        <span class="o_form_label"><strong><field name="lang" readonly="True" force_save="1"/></strong> has been successfully installed.
-Users can choose his favorite language in their preferences.</span>
-                    </group>
-                    <footer states="init">
+                    <footer>
                         <button name="lang_install" string="Add" type="object" class="btn-primary"/>
                         <button special="cancel" data-hotkey="z" string="Cancel" class="btn-secondary"/>
                     </footer>
-                    <footer states="done">
-                        <button name="reload" string="Close" type="object" class="btn-primary" data-hotkey="q"/>
-                        <button name="switch_lang" type="object" class="btn-primary" data-hotkey="w">Switch to <field name="lang" readonly="True" force_save="1"/> &amp; Close</button>
-                    </footer>
                 </form>
            </field>
         </record>
 
         <record id="action_view_base_language_install" model="ir.actions.act_window">
-            <field name="name">Add Language</field>
+            <field name="name">Add Languages</field>
             <field name="type">ir.actions.act_window</field>
             <field name="res_model">base.language.install</field>
             <field name="view_mode">form</field>
